### ISSUE TITLE
How to read continuous image frames for training?

### ISSUE THREAD

**Issue by MMagicLoren** [ultralytics/ultralytics#14565]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have a series of image frames of a moving target, but the speed of the target movement is unknown, I want to use YOLO to train this data and learn a motion model of the moving target. When given another untrained sequence, the motion target can be detected.
The image sequence is shown below：
![image](https://github.com/user-attachments/assets/2c5c812d-cc1a-45cf-9ce4-2c6a733397b1)
![image](https://github.com/user-attachments/assets/e854a7c1-1669-44b4-8801-ca2c2a799d0d)
![image](https://github.com/user-attachments/assets/f588fbc1-bb9c-4d6a-ac9e-da18e17f1b9a)
The targets in the images are highly similar and using motion information to detect the target is one way of detection.
Therefore, during the training phase of the network model, the training data reads must be continuous frames, how should I implement this?I will be grateful if you can give suggestions.

**Comment by Y-T-G** [comment]
This is not what YOLO is for. You should look at Kalman Filter, or combine YOLO predictions with CatBoost model.

**Comment by MMagicLoren** [comment]
Thank you for your reply. Although YOLO's main contribution does not lie in this type of detection, can its excellent detection performance be extended to this task? For Kalman filtering, certain known information is required, however, in this task, it is desired to detect moving targets without any known information.

**Comment by Y-T-G** [comment]
Like I said, you can train a CatBoost, LightGBM, or XGBoost model on the output of the YOLO model that takes in the last N detected coordinates of the object and outputs the next position.

**Comment by MMagicLoren** [comment]
Do you mean that YOLO detects all point targets in the image (whether they are moving or stationary) and feeds all the detected targets into a CatBoost model for training to do a classification problem, am I understanding this correctly?

**Comment by Y-T-G** [comment]
They can perform regression too.

**Comment by MMagicLoren** [comment]
Thanks again for your reply. I have one more doubt I hope I can get your help. the output of the YOLO model is the target prediction box and labels. Is it feeding these two data directly into CatBoost? If the labels predicted by YOLO are wrong, won't this lead to errors in the CatBoost trained model?

**Comment by Y-T-G** [comment]
You can use the ground truth labels. For inference, you use the YOLO predictions.

**Comment by MMagicLoren** [comment]
I used simulation to generate 100 sequences with 1-3 moving targets each, with moving targets labelled as 1 and stationary targets labelled as 0. The form of the data is shown below:
![image](https://github.com/user-attachments/assets/72c6544b-a2f3-46b4-bdee-aa79e7ddcf00)
The training code is as follows:
import os
import pandas as pd
from glob import glob
from sklearn.model_selection import train_test_split
from catboost import CatBoostClassifier, Pool


def read_single_file(file_path):
    """
    Reads a single Excel file and returns a DataFrame.
    """
    try:
        df = pd.read_excel(file_path)
        df['file_name'] = os.path.basename(file_path)
        return df
    except Exception as e:
        print(f"Error reading file {file_path}: {e}")
        return None


def read_all_trajectories(folder_path):
    """
    Reads all Excel files in a folder and returns a single DataFrame.
    """
    all_data = []
    excel_files = glob(os.path.join(folder_path, '*.xlsx'))

    for file_path in excel_files:
        df = read_single_file(file_path)
        if df is not None:
            all_data.append(df)

    if all_data:
        return pd.concat(all_data, ignore_index=True)
    else:
        return pd.DataFrame()

# read all trajectory data
folder_path = "./target_trajectories"
all_trajectory_data = read_all_trajectories(folder_path)

X = all_trajectory_data.drop(columns=['y', 'file_name'])
y = all_trajectory_data['y']
# split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# create a Pool object for training and testing
train_pool = Pool(X_train, y_train)
test_pool = Pool(X_test, y_test)
# create a model
model = CatBoostClassifier(task_type='GPU', learning_rate=0.03)
# train the model
model.fit(train_pool, eval_set=test_pool, use_best_model=True, plot=True)

best_loss = model.get_best_score()
acc = model.score(test_pool)
# save the model
model.save_model('catboost_model')

# load the model
load_model = CatBoostClassifier().load_model('catboost_model')
# read test data
test_data = pd.read_excel('./test_data/sequence_1.xlsx')
# predict
result = load_model.predict(test_data)
test_data['y'] = result
test_data.to_excel('./test_data/sequence_1.xlsx')
----------------------------------------------------
acc:0.9540788294149967 loss：0.16280853738324513
---------------------------------------------------------------------
Finally the test is performed, test_data has 3 motion targets but the predictions are all 0. I think the model is not learning the relationship between the frames, how can I improve it. I would appreciate if you can give suggestions. Looking forward to your reply.

**Comment by Y-T-G** [comment]
No. The features should be the last N detections for the same object. If N=3, you will have 6 features (3 * 2): `x1, y1, x2, y2, x3, y3`. 3 consecutive x and y coordinates of the same object. And the label or output will be x and y for the next frame. So two outputs.

**Comment by MMagicLoren** [comment]
Thanks for the suggestion, but I find that this requires the strong condition that the targets detected in consecutive frames must be able to correspond one to the other. Do you have any other good suggestions? How to make sure that the targets detected in N frames are associated one to one to construct features.

**Comment by Y-T-G** [comment]
You can use tracking.
